File: Stripe/StripeiOSTests/STPBankAccountParamsTest.m
//
//  STPBankAccountParamsTest.m
//  Stripe
//
//  Created by Joey Dong on 6/19/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@import XCTest;



@interface STPBankAccountParams ()

- (NSString *)accountHolderTypeString;

@end

@interface STPBankAccountParamsTest : XCTestCase

@end

@implementation STPBankAccountParamsTest

#pragma mark -

- (void)testLast4ReturnsAccountNumberLast4 {
    STPBankAccountParams *bankAccountParams = [[STPBankAccountParams alloc] init];
    bankAccountParams.accountNumber = @"000123456789";
    XCTAssertEqualObjects(bankAccountParams.last4, @"6789");
}

- (void)testLast4ReturnsNilWhenNoAccountNumberSet {
    STPBankAccountParams *bankAccountParams = [[STPBankAccountParams alloc] init];
    XCTAssertNil(bankAccountParams.last4);
}

- (void)testLast4ReturnsNilWhenAccountNumberIsLessThanLength4 {
    STPBankAccountParams *bankAccountParams = [[STPBankAccountParams alloc] init];
    bankAccountParams.accountNumber = @"123";
    XCTAssertNil(bankAccountParams.last4);
}

#pragma mark - STPBankAccountHolderType Tests

- (void)testAccountHolderTypeFromString {
    XCTAssertEqual([STPBankAccountParams accountHolderTypeFromString:@"individual"], STPBankAccountHolderTypeIndividual);
    XCTAssertEqual([STPBankAccountParams accountHolderTypeFromString:@"INDIVIDUAL"], STPBankAccountHolderTypeIndividual);

    XCTAssertEqual([STPBankAccountParams accountHolderTypeFromString:@"company"], STPBankAccountHolderTypeCompany);
    XCTAssertEqual([STPBankAccountParams accountHolderTypeFromString:@"COMPANY"], STPBankAccountHolderTypeCompany);

    XCTAssertEqual([STPBankAccountParams accountHolderTypeFromString:@"garbage"], STPBankAccountHolderTypeIndividual);
    XCTAssertEqual([STPBankAccountParams accountHolderTypeFromString:@"GARBAGE"], STPBankAccountHolderTypeIndividual);
}

- (void)testStringFromAccountHolderType {
    NSArray<NSNumber *> *values = @[
                                    @(STPBankAccountHolderTypeIndividual),
                                    @(STPBankAccountHolderTypeCompany),
                                    ];

    for (NSNumber *accountHolderTypeNumber in values) {
        STPBankAccountHolderType accountHolderType = (STPBankAccountHolderType)[accountHolderTypeNumber integerValue];
        NSString *string = [STPBankAccountParams stringFromAccountHolderType:accountHolderType];

        switch (accountHolderType) {
            case STPBankAccountHolderTypeIndividual:
                XCTAssertEqualObjects(string, @"individual");
                break;
            case STPBankAccountHolderTypeCompany:
                XCTAssertEqualObjects(string, @"company");
                break;
        }
    }
}

#pragma mark - Description Tests

- (void)testDescription {
    STPBankAccountParams *bankAccountParams = [[STPBankAccountParams alloc] init];
    XCTAssert(bankAccountParams.description);
}

#pragma mark - STPFormEncodable Tests

- (void)testRootObjectName {
    XCTAssertEqualObjects([STPBankAccountParams rootObjectName], @"bank_account");
}

- (void)testPropertyNamesToFormFieldNamesMapping {
    STPBankAccountParams *bankAccountParams = [[STPBankAccountParams alloc] init];

    NSDictionary *mapping = [STPBankAccountParams propertyNamesToFormFieldNamesMapping];

    for (NSString *propertyName in [mapping allKeys]) {
        XCTAssertFalse([propertyName containsString:@":"]);
        XCTAssert([bankAccountParams respondsToSelector:NSSelectorFromString(propertyName)]);
    }

    for (NSString *formFieldName in [mapping allValues]) {
        XCTAssert([formFieldName isKindOfClass:[NSString class]]);
        XCTAssert([formFieldName length] > 0);
    }

    XCTAssertEqual([[mapping allValues] count], [[NSSet setWithArray:[mapping allValues]] count]);
}

- (void)testAccountHolderTypeString {
    STPBankAccountParams *bankAccountParams = [[STPBankAccountParams alloc] init];

    bankAccountParams.accountHolderType = STPBankAccountHolderTypeIndividual;
    XCTAssertEqualObjects([bankAccountParams accountHolderTypeString], @"individual");

    bankAccountParams.accountHolderType = STPBankAccountHolderTypeCompany;
    XCTAssertEqualObjects([bankAccountParams accountHolderTypeString], @"company");
}

@end

File: Stripe/StripeiOSTests/STPBankAccountParamsTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPBankAccountParamsTest.m
//  Stripe
//
//  Created by Joey Dong on 6/19/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@testable import Stripe
@testable import StripePayments
import XCTest

class STPBankAccountParamsTest: XCTestCase {
    // MARK: -

    func testLast4ReturnsAccountNumberLast4() {
        let bankAccountParams = STPBankAccountParams()
        bankAccountParams.accountNumber = "000123456789"
        XCTAssertEqual(bankAccountParams.last4, "6789")
    }

    func testLast4ReturnsNilWhenNoAccountNumberSet() {
        let bankAccountParams = STPBankAccountParams()
        XCTAssertNil(bankAccountParams.last4)
    }

    func testLast4ReturnsNilWhenAccountNumberIsLessThanLength4() {
        let bankAccountParams = STPBankAccountParams()
        bankAccountParams.accountNumber = "123"
        XCTAssertNil(bankAccountParams.last4)
    }

    // MARK: - STPBankAccountHolderType Tests

    func testAccountHolderTypeFromString() {
        XCTAssertEqual(STPBankAccountParams.accountHolderType(from: "individual"), STPBankAccountHolderType.individual)
        XCTAssertEqual(STPBankAccountParams.accountHolderType(from: "INDIVIDUAL"), STPBankAccountHolderType.individual)

        XCTAssertEqual(STPBankAccountParams.accountHolderType(from: "company"), STPBankAccountHolderType.company)
        XCTAssertEqual(STPBankAccountParams.accountHolderType(from: "COMPANY"), STPBankAccountHolderType.company)

        XCTAssertEqual(STPBankAccountParams.accountHolderType(from: "garbage"), STPBankAccountHolderType.individual)
        XCTAssertEqual(STPBankAccountParams.accountHolderType(from: "GARBAGE"), STPBankAccountHolderType.individual)
    }

    func testStringFromAccountHolderType() {
        let values = [
            STPBankAccountHolderType.individual,
            STPBankAccountHolderType.company,
        ]

        for accountHolderType in values {
            let string = STPBankAccountParams.string(from: accountHolderType)

            switch accountHolderType {
            case STPBankAccountHolderType.individual:
                XCTAssertEqual(string, "individual")
            case STPBankAccountHolderType.company:
                XCTAssertEqual(string, "company")
            default:
                break
            }
        }
    }

    // MARK: - Description Tests

    func testDescription() {
        let bankAccountParams = STPBankAccountParams()
        XCTAssertNotNil(bankAccountParams.description)
    }

    // MARK: - STPFormEncodable Tests

    func testRootObjectName() {
        XCTAssertEqual(STPBankAccountParams.rootObjectName(), "bank_account")
    }

    func testPropertyNamesToFormFieldNamesMapping() {
        let bankAccountParams = STPBankAccountParams()

        let mapping = STPBankAccountParams.propertyNamesToFormFieldNamesMapping()

        for propertyName in mapping.keys {
            XCTAssertFalse(propertyName.contains(":"))
            XCTAssert(bankAccountParams.responds(to: NSSelectorFromString(propertyName)))
        }

        for formFieldName in mapping.values {
            XCTAssert(!formFieldName.isEmpty)
        }

        XCTAssertEqual(mapping.values.count, Set(mapping.values).count)
    }

    func testAccountHolderTypeString() {
        let bankAccountParams = STPBankAccountParams()

        bankAccountParams.accountHolderType = STPBankAccountHolderType.individual
        XCTAssertEqual(bankAccountParams.accountHolderTypeString(), "individual")

        bankAccountParams.accountHolderType = .company
        XCTAssertEqual(bankAccountParams.accountHolderTypeString(), "company")
    }
}
